File: http_proxy/shared/policy.py
"""Policy engine. Loads rules from YAML, evaluates requests."""

from dataclasses import dataclass
from enum import Enum
from pathlib import Path

import yaml


class Action(Enum):
    ALLOW = "allow"
    DENY = "deny"
    APPROVAL = "approval"


@dataclass(frozen=True)
class Decision:
    action: Action
    reason: str


@dataclass(frozen=True)
class Rule:
    match: dict[str, str]
    action: Action
    label: str = ""


def load_rules(path: Path) -> list[Rule]:
    """Load policy rules from YAML file."""
    data = yaml.safe_load(path.read_text())
    rules = []
    for entry in data.get("rules", []):
        rules.append(Rule(
            match=entry["match"],
            action=Action(entry["action"]),
            label=entry.get("label", ""),
        ))
    return rules


def evaluate(fields: dict[str, str], rules: list[Rule]) -> Decision:
    """Evaluate request fields against rules. First match wins. Default deny."""
    for rule in rules:
        if _matches(fields, rule.match):
            return Decision(
                action=rule.action,
                reason=rule.label or str(rule.match),
            )
    return Decision(action=Action.DENY, reason="no matching rule")


def _matches(fields: dict[str, str], match: dict[str, str]) -> bool:
    """All match conditions must be satisfied."""
    for key, pattern in match.items():
        value = fields.get(key, "")
        if pattern == "*":
            continue
        if key.endswith("_contains"):
            actual_key = key.removesuffix("_contains")
            if pattern not in fields.get(actual_key, ""):
                return False
        elif value != pattern:
            return False
    return True

File: mcp_gateway/src/policy.ts
import type { RuleConfig } from "./config.js";

export enum Action {
  ALLOW = "allow",
  DENY = "deny",
  APPROVAL = "approval",
}

export interface Decision {
  action: Action;
  reason: string;
}

export interface Rule {
  match: Record<string, string>;
  action: Action;
  label: string;
}

export function parseRules(configs: RuleConfig[]): Rule[] {
  return configs.map((c) => ({
    match: c.match,
    action: c.action as Action,
    label: c.label ?? "",
  }));
}

export function evaluate(
  fields: Record<string, string>,
  rules: Rule[]
): Decision {
  for (const rule of rules) {
    if (matches(fields, rule.match)) {
      return { action: rule.action, reason: rule.label || JSON.stringify(rule.match) };
    }
  }
  return { action: Action.DENY, reason: "no matching rule" };
}

function matches(
  fields: Record<string, string>,
  match: Record<string, string>
): boolean {
  for (const [key, pattern] of Object.entries(match)) {
    if (pattern === "*") continue;

    if (key.endsWith("_contains")) {
      const actualKey = key.slice(0, -"_contains".length);
      const value = fields[actualKey] ?? "";
      if (!value.includes(pattern)) return false;
    } else {
      const value = fields[key] ?? "";
      if (value !== pattern) return false;
    }
  }
  return true;
}
